Add docs for ROS integration and ROSEnv generator (#3891)
conan-io/docs: integrations/ros.rst

.. _integrations_ros:

|ros_logo| ROS
==============

.. include:: ../common/experimental_warning.inc

Conan provides integration for your Robot Operating System (ROS) C/C++ based projects.
This will allow you to consume Conan packages inside your ROS package projects.

> The Conan packages can be installed and used in CMake with the help of the :ref:`ROSEnv generator <conan_tools_ros_rosenv>` generator created

Reviewer: ```suggestion
The Conan packages can be installed and used in CMake with the help of the :ref:`ROSEnv generator <conan_tools_ros_rosenv>` created
```

for the purpose.

> It provides a clean integration that requires no conan-specific modifications in your *CMakeLists.txt*.

Reviewer: ```suggestion
It provides a clean integration that requires no Conan-specific modifications in your *CMakeLists.txt*.
```

> .. note::

Reviewer: I think it would be nice to add some pre-requisites to run the example, like for example run everything inside the `osrf/ros:humble-desktop` docker image to test it.

Author: done

    The files for this example can be found in `this repository <https://github.com/conan-io/examples2/tree/main/examples/tools/ros/rosenv/workspace>`_.
    You can follow this example by cloning our examples repository:

    .. code-block:: bash

        $ git clone git@github.com:conan-io/examples2.git && cd examples2/examples/tools/ros/rosenv

.. important::

    This integration supports ROS2 and has been developed using ROS Humble version and the aim is to support newer versions going forward.

>     If you have any issues with other ROS versions, please let us know by opening an issue in our github repository.

Reviewer: ```suggestion
    If you have any issues with other ROS versions, please let us know by opening an issue in our GitHub repository.
```

Consuming Conan packages using the ROSEnv generator
---------------------------------------------------

> Imagine we have a ROS C++ pacakge called `str_printer` that uses some functionality from the 

Reviewer: ```suggestion
Imagine we have a ROS C++ package called `str_printer` that uses some functionality from the 
```

> Imagine we have a ROS C++ package called `str_printer` that uses some functionality from the 
> third party library `fmt <https://conan.io/center/recipes/fmt>`_ a `str_printer` library with a function to print fancy strings.

Reviewer: ```suggestion
Imagine we have a ROS C++ package called `str_printer` that uses some functionality from the 
third party string formatting library `fmt <https://conan.io/center/recipes/fmt>`_ to print fancy strings.
```

We have the following project structure:

.. code-block:: bash

    $  tree /f
    workspace
    ├───str_printer
    │   │   CMakeLists.txt
    │   │   conanfile.txt
    │   │   package.xml
    │   ├───include
    │   │   └──str_printer
    │   │         str_printer.h
    │   └───src
    │          str_printer.cpp
    └───consumer
        │   CMakeLists.txt
        │   package.xml
        └───src
               main.cpp

Where:

- The `str_printer` is a ROS package that implements a function and **depends on the fmt Conan package**.

> - The `consumer` is also a ROS package that dependes on the `str_printer` ROS package and uses its functionality in a **final executable**.

Reviewer: ```suggestion
- The `consumer` is also a ROS package that depends on the `str_printer` ROS package and uses its functionality in a **final executable**.
```

    conanfile.txt: Generator 'CMakeToolchain' calling 'generate()'
    conanfile.txt: CMakeToolchain generated: conan_toolchain.cmake
    conanfile.txt: Preset 'conan-release' added to CMakePresets.json. Invoke it manually using 'cmake --preset conan-release' if using CMake>=3.23
    conanfile.txt: If your CMake version is not compatible with CMakePresets (<3.23) call cmake like: 'cmake <path> -G "Unix Makefiles" -DCMAKE_TOOLCHAIN_FILE=/home/danimtb/examples2/examples/tools/ros/rosenv/workspace/install/conan/conan_toolchain.cmake -DCMAKE_POLICY_DEFAULT_CMP0091=NEW -DCMAKE_BUILD_TYPE=Release'
    conanfile.txt: CMakeToolchain generated: CMakePresets.json
    conanfile.txt: CMakeToolchain generated: ../../str_printer/CMakeUserPresets.json
    conanfile.txt: Generator 'ROSEnv' calling 'generate()'
    conanfile.txt: Generated ROSEnv Conan file: conanrosenv.sh
    Use 'source /home/user/examples2/examples/tools/ros/rosenv/workspace/install/conan/conanrosenv.sh' to set the ROSEnv Conan before 'colcon build'
    conanfile.txt: Generating aggregated env files
    conanfile.txt: Generated aggregated env files: ['conanrosenv.sh']
    Install finished successfully

> This will download the fmt Conan package to the local cache and generate the CMake and ROS environment files

Reviewer: ```suggestion
This will download the `fmt` Conan package to the local cache and generate the CMake and ROS environment files
```

in the `conan` subfolder of the `install` directory.

> Now we can source our ROS environment, then **source the Conan ROSEnv environment**, so the conan-installed pacakges are found by CMake, and then

Reviewer: ```suggestion
Now we can source our ROS environment, then **source the Conan ROSEnv environment**, so the conan-installed package are found by CMake, and then
```

.. code-block:: text
   :caption: str_printer/CMakeLists.txt
   :emphasize-lines: 10,22

    cmake_minimum_required(VERSION 3.8)
    project(str_printer)

    if(CMAKE_COMPILER_IS_GNUCXX OR CMAKE_CXX_COMPILER_ID MATCHES "Clang")
    add_compile_options(-Wall -Wextra -Wpedantic)
    endif()

    # find dependencies
    find_package(ament_cmake REQUIRED)
    find_package(fmt REQUIRED)  # Retrieved with Conan C/C++ Package Manager

    add_library(str_printer src/str_printer.cpp)

    target_include_directories(str_printer PUBLIC
    $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include/str_printer>
    $<INSTALL_INTERFACE:include>)

    target_compile_features(str_printer PUBLIC c_std_99 cxx_std_17)  # Require C99 and C++17
    ament_target_dependencies(str_printer fmt)

    ament_export_targets(str_printerTargets HAS_LIBRARY_TARGET)
    ament_export_dependencies(fmt)

    install(
    DIRECTORY include/
    DESTINATION include
    )

    install(
    TARGETS str_printer
    EXPORT str_printerTargets
    LIBRARY DESTINATION lib
    ARCHIVE DESTINATION lib
    RUNTIME DESTINATION bin
    INCLUDES DESTINATION include
    )

    ament_package()

> To build the `consumer` ROS pacakge, you can proceed as usual (make sure

Reviewer: ```suggestion
To build the `consumer` ROS package, you can proceed as usual (make sure
```
